kale/pipeline/multi_domain_adapter.py

```diff
@@ -0,0 +1,314 @@
+            if domain_ != target_domain:
+                self.classifiers[domain_] = task_classifier(self.feature_dim, n_classes)
+        # init classifiers as nn.ModuleDict, otherwise it will not be optimized
+        self.classifiers = nn.ModuleDict(self.classifiers)
+        self.k_moment = k_moment
+
+    def compute_loss(self, batch, split_name="V"):
+        x, y, domain_labels = batch
+        phi_x = self.forward(x)
+        moment_loss = self._compute_domain_dist(phi_x, domain_labels)
+        src_idx = torch.where(domain_labels != self.target_label)
+        tgt_idx = torch.where(domain_labels == self.target_label)
+        cls_loss, ok_src = self._compute_cls_loss(phi_x[src_idx], y[src_idx], domain_labels[src_idx])
+        if len(tgt_idx) > 0:
+            y_tgt_hat = _average_cls_output(phi_x[tgt_idx], self.classifiers)
+            _, ok_tgt = losses.cross_entropy_logits(y_tgt_hat, y[tgt_idx])
+        else:
+            ok_tgt = 0.0
+
+        task_loss = cls_loss
+        log_metrics = {
+            f"{split_name}_source_acc": ok_src,
+            f"{split_name}_target_acc": ok_tgt,
+            f"{split_name}_domain_acc": moment_loss,
+        }
+
+        return task_loss, moment_loss, log_metrics
+
+    def _compute_cls_loss(self, x, y, domain_labels: torch.Tensor):
+        if len(y) == 0:
+            return 0.0, 0.0
+        else:
+            cls_loss = 0.0
+            ok_src = []
+            n_src = 0
+            for domain_ in self.domain_to_idx.keys():
+                if domain_ == self.target_domain:
+                    continue
+                domain_idx = torch.where(domain_labels == self.domain_to_idx[domain_])
+                cls_output = self.classifiers[domain_](x[domain_idx])
+                loss_cls_, ok_src_ = losses.cross_entropy_logits(cls_output, y[domain_idx])
+                cls_loss += loss_cls_
+                ok_src.append(ok_src_)
+                n_src += 1
+            cls_loss = cls_loss / n_src
+            ok_src = torch.cat(ok_src)
+            return cls_loss, ok_src
+
+    def _compute_domain_dist(self, x, domain_labels):
+        """Compute k-th order moment divergence
+
+        Args:
+            x (torch.Tensor): input data, shape (n_samples, n_features)
+            domain_labels (torch.Tensor): labels indicate the instance from which domain, shape (n_samples,)
+
+        Returns:
+            torch.Tensor: divergence
+        """
+
+        moment_loss = 0
+        for i in range(self.k_moment):
+            moment_loss += losses._moment_k(x, domain_labels, i + 1)
+
+        return moment_loss
+
+
+class _DINTrainer(BaseMultiSourceTrainer):
+    def __init__(
+        self,
+        dataset,
+        feature_extractor,
+        task_classifier,
+        n_classes: int,
+        target_domain: str,
+        kernel: str = "linear",
+        kernel_mul: float = 2.0,
+        kernel_num: int = 5,
+        **base_params,
+    ):
+        """Domain independent network.
```

Comment:
No further explanation? 
Is this a published/existing work? 
If it is new, just say this is a new network in development that will be updated with references later.